src/main/ruby/core/posix.rb

```diff
@@ -292,11 +296,9 @@ def self.write_string(fd, string, nonblock=false)
     while written < length
       ret = write(fd, buffer + written, length - written)
       if ret < 0
-        if nonblock
-          errno = Errno.errno
-          if errno == Errno::EAGAIN::Errno || errno == Errno::EWOULDBLOCK::Errno
-            raise IO::EAGAINWaitWritable
-          end
+        errno = Errno.errno
+        if errno == Errno::EAGAIN::Errno || errno == Errno::EWOULDBLOCK::Errno
+          raise IO::EAGAINWaitWritable
         end
```

Comment:
This is problematic as mentioned in the comment below.
If we get EAGAIN after writing a few bytes, the caller cannot know how much we wrote, and we will end up writing the same data twice or more times.

It seems a nice way to handle this is to move the `select/retry` logic here if blocking==true.